tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+  $host_device_id =~ /pci_([a-z\d]+)_([a-z\d]+)_([a-z\d]+)_([a-z\d]+)/;
+  my ($dom, $bus, $slot, $func) = ($1, $2, $3, $4);  #these are different with those in host_devices.xml 
+  for (my $i=0; $i<$nics_count; $i++) {
+    $devs_xml =~ /\<interface.*?type='hostdev'.*?\<\/interface\>/s;    #minimal match. 
+    my $nic_xml = $&;
+    if ($nic_xml =~ /domain='(0x)?$dom' +bus='(0x)?$bus' +slot='(0x)?$slot' +function='(0x)?$func'/){
+
+      #get the vf xml in vm config file for later unplugging to use
+      $nic{xml} = $nic_xml;
+
+      #get mac address
+      $nic_xml =~ /\<mac.*address='(.*)'/;
+      $nic{mac} = $1;
+
+      #get bdf for guests on KVM
+      $nic_xml =~ s/\<source.*\<\/source\>//s;
+      if ($nic_xml =~ /bus='(\w+)'.*slot='(\w+)'.*function='(\w+)'/) {
+        my ($bus, $slot, $func) = ($1, $2, $3);
+        $nic{bdf} = $bus . ":" . $slot . "." . $func;
+      }
+      else {
+        #have to get bdf by other means for guests on XEN
+        #but no these file on pv guest on XEN, need to find another way!! ???
+        $nic{bdf} = script_output "ssh root\@$vm \"if [ -e /sys/devices/pci-0/pci????:?? ]; then grep -H '$nic{mac}' /sys/devices/pci-0/*/*/net/*/address | cut -d '/' -f6; else grep -H '$nic{mac}' /sys/devices/*/*/net/*/address | cut -d '/' -f5; fi\"";
+#        $nic{bdf} = script_output "ssh root\@$vm \"grep -H '$nic{mac}' /sys/devices/*/*/net/*/address | cut -d '/' -f5\"";
+      }
+      last;
+
+    }
+    else {
+      $devs_xml =~ s/\<interface.*?type='hostdev'.*?\<\/interface\>//s;  
+    } 
+  }
+
+  #print the vf device and return
+  if ( $nic{bdf} =~ /[a-z\d]+:[a-z\d]+[.:][a-z\d]+/ ) {    #bdf has different format in guests on KVM and XEN
+    script_run "ssh root\@$vm \"lspci -vvv -s $nic{bdf}\"";
+    save_screenshot;
+    record_info("VF plugged to vm", "$host_device_id \nGuest: $vm\nbdf='$nic{bdf}'   mac_addrss='$nic{mac}'");
+    return %nic;
+  }
+
+}
+
+
+#unplug the vf device from guest
+sub unplug_vf_from_vm {
+  my ($vm, $mac, $device_xml) = @_;
+
+  #bring the nic down
+  my $nic = script_output "ssh root\@$vm \"grep '$mac' /sys/class/net/*/address | cut -d'/' -f5 | head -n1\"";
+  script_run("ssh root\@$vm 'ifdown $nic'", 60);
+
+  #detach vf from guest
+  my $vf_xml_file = "vf_in_vm.xml";
+  assert_script_run "echo \"$device_xml\" > $vf_xml_file";
+  assert_script_run("virsh detach-device $vm $vf_xml_file --persistent", 60);
+
+}
+
+
+#check host and guest to make sure they work well
+sub check_host_and_guest() { 
+  my ($self, $vm) = @_;
+
+  #check if guest is still alive
+  if (script_output("virsh domstate $vm", 30) ne "running") {;  #need check status ???
+    $self->{test_results}->{$vm}->{"Error: Please check vm status as $vm is not running!"}->{status} = 'FAILED';    
+    script_run "virsh list --all";
+    return 1;
+  }
+
+  #other checks like checking journals from guest???
+  #need check the oops bug
+
```

Comment:
You can implement those are really wanted for sr-iov here. If you think there are other things to be checked and, at the  same time general enough to be shared with others, please implement a standalone function in other places.

Reply:
Yes, agree.